### ISSUE TITLE
Training on a custom dataset

### ISSUE THREAD

**Issue by shashankg7** [zhangxiangxiao/Crepe#21]
Hi Zhang, 

Thanks for open sourcing this code. I am trying to run your model on my dataset, I have pre-processed the data in it's required format. 

Now I am not familiar with lua and torch, so can you tell me is there anything I need to change in the config file to run it over a different dataset than dbpedia one? (I could only decipher the no of classes parameter from config file). Sorry if the question sounds elementary, I just want to make sure I am folllowing the right steps for my experiements.

Thanks in advance.

**Comment by sivagnanamn** [comment]
All that you should change is the number of output classes in the config file. In the below line, change the outputSize parameter to the number of output classes.

config.model[23] = {module = "nn.Linear", inputSize = 1024, outputSize = #NoOfOutputClasses}

**Comment by shashankg7** [comment]
Thanks for your reply. I tried that and it's working. Also I have one more small question. 

I am getting accuracy of around 18% after 20000 epochs. Task is 8 class classification and training data has around 65k samples and testing data has 16k samples. 

Can you think of something from top of your head which I might have done wrong?

**Comment by sivagnanamn** [comment]
Your question is too generic. It totally depends on your data. First thing that I can think of is 
- Check whether data is balanced
- You can try some pre processing to remove unwanted info from the data

Please close the ticket if your query is resolved.

**Comment by shashankg7** [comment]
Thanks for your response. I realized that maximum length it is taking is around 2500 which will consume more memory per batch. 

I trimmed it to avg. sentence length instead of max. sentence length which is ~140 so now the code is running in reasonable time.

**Comment by shashankg7** [comment]
Sorry I closed it by mistake.

Classes are imbalanced. Half of the classes have most of the documents, but baseline methods (logistic regression on top of wordvectors) is performing far better than this (~62%).

That was the reason I was worried about.

**Comment by zhangxiangxiao** [comment]
If your dataset is imbalanced as you said, assigning all samples to the largest class can give you around 50% accuracy. 62% is not surprising at all.

However, the Crepe code enforces **balanced sampling from classes** at training, because of [line 65 in data.lua](https://github.com/zhangxiangxiao/Crepe/blob/master/train/data.lua#L65). It does not enforce such balanced sampling during testing. The mismatch between training and testing distributions will result in funny results.

**Comment by shashankg7** [comment]
It's a multi-class (8 class) classification problem, so how will assigning all samples to largest class will give 50% acc.?

**Comment by shashankg7** [comment]
Maybe I was wrong in my statement. I overestimated the class imbalance. Here is the distribution of classes :

7: 23849, 0: 15159, 1: 6445, 4: 5759, 5: 3969, 3: 3659, 2: 2845, 6: 492

I am sure 62% is significant considering this distribution.

**Comment by zhangxiangxiao** [comment]
You statistics also show a very small data size (in terms of number of samples). It would be very hard for convolutional networks to learn good representation in that case.

I will close the issue now because the topic start to deviate away from your original question. Feel free to open a new issue for different questions.

**Comment by shashankg7** [comment]
Sure, thanks for your reply. I was just curious since other models (simpler models) were performing better than this one.

**Comment by zhangxiangxiao** [comment]
Bad performance is expected if your dataset is less than several hundred thousand samples. If you are interested, refer to our article for a set of complete benchmarks.

**Comment by shashankg7** [comment]
Sure, thanks for your quick response. Will refer to your article. Thanks again.
